File: scripts/generate_sitemap.py
"""
Karnata.in — generate_sitemap.py
Generates a comprehensive, valid sitemap.xml for Google Search Console, Bing Webmaster,
and AI Search Engines (GEO - ChatGPT, Perplexity, Google SGE, Claude).
"""

import os
import sys
from pathlib import Path
from datetime import datetime

if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')

BASE_DIR = Path(__file__).parent.parent
SITEMAP_PATH = BASE_DIR / "sitemap.xml"
BASE_URL = "https://karnata.in"
TODAY = datetime.now().strftime("%Y-%m-%d")

NOINDEX_DIRS = {'admin', 'cms', 'studio', 'imd_hub', 'templates', 'scratch', '.git', 'node_modules', '.wrangler'}
NOINDEX_FILES = {
    'admin.html', 'admin-transfers.html', 'admin-news.html', 'admin-login.html',
    'cms-admin.html', '404.html', 'error.html', 'test.html'
}

def generate_sitemap():
    entries = []
    seen_urls = set()

    for root, dirs, files in os.walk(BASE_DIR):
        rel_dir = Path(root).relative_to(BASE_DIR).as_posix()
        if any(part in NOINDEX_DIRS for part in rel_dir.split('/')):
            continue
        for f in files:
            if f.endswith('.html') and f not in NOINDEX_FILES and 'admin' not in f and not f.startswith('_'):
                f_path = Path(root) / f
                rel_p = f_path.relative_to(BASE_DIR).as_posix()
                
                url = f"{BASE_URL}/{rel_p}"
                if rel_p == "index.html":
                    url = f"{BASE_URL}/"

                if url not in seen_urls:
                    seen_urls.add(url)
                    prio = "1.0" if url == f"{BASE_URL}/" else ("0.9" if any(k in url for k in ['officers', 'news', 'gold', 'dam', 'districts', 'jyothishya']) else "0.8")
                    freq = "hourly" if any(k in url for k in ['news', 'gold']) else "daily"
                    entries.append({
                        "loc": url,
                        "priority": prio,
                        "changefreq": freq
                    })

    # Sort entries: Homepage first, then high priority
    entries.sort(key=lambda x: (0 if x["loc"] == f"{BASE_URL}/" else 1, -float(x["priority"]), x["loc"]))

    # Build XML
    xml_lines = [
        '<?xml version="1.0" encoding="UTF-8"?>',
        '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"',
        '        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"',
        '        xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9',
        '        http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">'
    ]

    for item in entries:
        xml_lines.append('  <url>')
        xml_lines.append(f'    <loc>{item["loc"]}</loc>')
        xml_lines.append(f'    <lastmod>{TODAY}</lastmod>')
        xml_lines.append(f'    <changefreq>{item["changefreq"]}</changefreq>')
        xml_lines.append(f'    <priority>{item["priority"]}</priority>')
        xml_lines.append('  </url>')

    xml_lines.append('</urlset>\n')

    sitemap_content = '\n'.join(xml_lines)
    SITEMAP_PATH.write_text(sitemap_content, encoding='utf-8')
    print(f"SUCCESS: Sitemap generated successfully with {len(entries)} URLs at {SITEMAP_PATH} (lastmod: {TODAY})")

if __name__ == "__main__":
    generate_sitemap()

File: scripts/generate_sitemap.js
const fs = require('fs');
const path = require('path');

const ROOT_DIR = path.join(__dirname, '..');
const SITEMAP_PATH = path.join(ROOT_DIR, 'sitemap.xml');
const ROOT_SITEMAP_PATH = path.join(path.dirname(ROOT_DIR), 'sitemap.xml');
const BASE_URL = 'https://karnata.in';
const TODAY = new Date().toISOString().split('T')[0];

const EXCLUDE_DIRS = new Set([
  'node_modules', '.git', 'scratch', 'logs', '.system_generated',
  '__pycache__', '.wrangler', '.vscode', 'admin', 'cms', 'imd_hub',
  'templates', 'studio', 'lib', 'scraper', 'functions'
]);

const EXCLUDE_FILES = new Set([
  'constituency-detail.html',
  'dam-details.html',
  'scheme-detail.html',
  'article.html',
  'news/article.html',
  'mla/index.html'
]);

function getPriority(relPath) {
  if (relPath === '' || relPath === '/') return '1.0';
  if (relPath.includes('privacy') || relPath.includes('terms') || relPath.includes('disclaimer')) return '0.7';
  if (relPath.startsWith('/article/') || relPath.startsWith('/news/') || relPath.includes('voter-roll') || relPath.includes('local-news')) return '1.0';
  if (relPath.startsWith('/districts/') || relPath.startsWith('/mla/') || relPath.startsWith('/mp/') || relPath.startsWith('/dam-levels/')) return '0.9';
  return '0.85';
}

function generateSitemap() {
  const urls = [];
  const seen = new Set();

  function add(p) {
    let clean = p.replace(/\\/g, '/');
    if (clean.endsWith('.html')) clean = clean.slice(0, -5);
    if (clean.endsWith('/index')) clean = clean.slice(0, -6);
    const cleanPath = clean.startsWith('/') ? clean : (clean ? '/' + clean : '');
    
    // Ignore numerical IDs, internal tests, or duplicate double 'aa' filenames
    if (/^\/mla\/\d+$/.test(cleanPath)) return;
    if (cleanPath.endsWith('aa_assembly_constituency')) return;
    if (cleanPath.includes('/templates/') || cleanPath.includes('/imd_hub/')) return;

    if (!seen.has(cleanPath)) {
      seen.add(cleanPath);
      urls.push({
        loc: BASE_URL + (cleanPath === '' ? '/' : cleanPath),
        priority: getPriority(cleanPath),
        changefreq: 'always',
        lastmod: TODAY
      });
    }
  }

  // Always ensure root is first
  add('');

  // Recursively scan all valid public HTML files in project
  function scanDir(dir) {
    if (!fs.existsSync(dir)) return;
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.name.startsWith('.')) continue;
      const fullPath = path.join(dir, entry.name);
      const relPath = path.relative(ROOT_DIR, fullPath).replace(/\\/g, '/');

      if (entry.isDirectory()) {
        if (!EXCLUDE_DIRS.has(entry.name) && !relPath.includes('namma-karnataka')) {
          scanDir(fullPath);
        }
      } else if (entry.isFile() && entry.name.endsWith('.html')) {
        if (!EXCLUDE_FILES.has(entry.name) && !EXCLUDE_FILES.has(relPath)) {
          add('/' + relPath);
        }
      }
    }
  }

  scanDir(ROOT_DIR);

  let xml = '<?xml version="1.0" encoding="UTF-8"?>\n';
  xml += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n';
  urls.forEach(u => {
    xml += '  <url>\n';
    xml += '    <loc>' + u.loc + '</loc>\n';
    xml += '    <lastmod>' + u.lastmod + '</lastmod>\n';
    xml += '    <changefreq>always</changefreq>\n';
    xml += '    <priority>' + u.priority + '</priority>\n';
    xml += '  </url>\n';
  });
  xml += '</urlset>\n';

  fs.writeFileSync(SITEMAP_PATH, xml, 'utf-8');
  if (fs.existsSync(path.dirname(ROOT_SITEMAP_PATH))) {
    try { fs.writeFileSync(ROOT_SITEMAP_PATH, xml, 'utf-8'); } catch(e) {}
  }
  console.log('✓ Generated clean sitemap.xml with dynamic auto-discovery across all ' + urls.length + ' URLs.');
}

generateSitemap();
